### PR TITLE
Devel

### PR DESCRIPTION
**Pull request by mhariharasudan**
libbarrett's version 1.2.4.
- Support for left and right proficio in the examples were included.
- Config files can be loaded from the ~/.barrett directory.
- Readme is updated to reflect the changes.

### CODE REVIEW

**Comment by mhariharasudan** [comment]
This version of libbarrett was tested in a new computer. The Readme is updated to reflect the new changes.

File: .bash_aliases [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# for Proficio
```
Comment (bzenowich) [inline .bash_aliases:1]: Readme should instruct user to cat .aliases >> ~/.bash_aliases && source it, instead of potentially overwriting their existing one.

File: .bash_aliases [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+alias leftConfig='cp ~/libbarrett/proficio_sandbox/configurations/wam3.conf.left /etc/barrett/wam3.conf; cp ~/libbarrett/proficio_sandbox/configurations/zerocal.conf.left /etc/barrett/calibration_data/wam3/zerocal.conf; cp ~/libbarrett/proficio_sandbox/configurations/calibration.conf.left /etc/barrett/calibration.conf; cp ~/libbarrett/proficio_sandbox/configurations/gravitycal.conf.left /etc/barrett/calibration_data/wam3/gravitycal.conf'
```
Comment (bzenowich) [inline .bash_aliases:2]: These .left and .right files should be in /etc/barrett, not ~/libbarrett. We shouldn't depend on the user checking out the code into ~/libbarrett.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,12 +1,9 @@
 cmake_minimum_required(VERSION 2.6)
 
 project(libbarrett)
-
-set(libbarrett_VERSION_MAJOR 1)
-set(libbarrett_VERSION_MINOR 3)
 set(libbarrett_VERSION_MAJOR 1)
 set(libbarrett_VERSION_MINOR 2)
-set(libbarrett_VERSION_PATCH 3)
+set(libbarrett_VERSION_PATCH 4)
```
Comment (bzenowich) [inline CMakeLists.txt:6]: We need to update support.barrett.com/Libbarrett and the main page with the new version info, and the new installation instructions.

Comment (bzenowich) [inline CMakeLists.txt:6]: We need to re-make the doxy files (html) and post them at web.barrett.com/libbarrett.

File: CMakeLists.txt [hunk 4]
```diff
@@ -75,6 +72,7 @@ set(exported_libraries "")
 
 ## Put libbarrett headers in the include path
 include_directories("${PROJECT_SOURCE_DIR}/include/")
+include_directories("${PROJECT_BINARY_DIR}/include")
```
Comment (bzenowich) [inline CMakeLists.txt:75]: What is the significance of the trailing "/" versus no slash on the include directory?

File: CMakeLists.txt [hunk 5]
```diff
@@ -236,11 +242,7 @@ configure_file(${PROJECT_SOURCE_DIR}/Doxyfile.in ${PROJECT_SOURCE_DIR}/Doxyfile
 	ESCAPE_QUOTES @ONLY
 )
 
-if (NON_REALTIME)
```
Comment (bzenowich) [inline CMakeLists.txt:239]: Why was this removed?

File: README.txt [hunk 6]
```diff
@@ -4,14 +4,54 @@ Barrett Technology
...
+$ tar -xf libconfig-1.4.5-PATCHED.tar.gz
+$ cd libconfig-1.4.5
+$ ./configure && make && sudo make install
+$ cd ../
+$ rm -rf libconfig-1.4.5 libconfig-1.4.5-PATCHED.tar.gz
+
+Download and install libbarrett:
+
+$ cd ~/
+$ git clone https://github.com/BarrettTechnology/libbarrett.git
```
Comment (bzenowich) [inline README.txt:30]: I think we need a "git checkout" command here to specify a particular branch, else it may refer to origin/master by default.

File: proficio_sandbox/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+
+function(add_programs)
+	foreach (prog ${ARGV})
+		add_executable(${prog} ${prog}.cpp)
+		target_link_libraries(${prog} ${BARRETT_LIBRARIES})
+	endforeach()
+endfunction()
+
+add_programs(
```
Comment (bzenowich) [inline proficio_sandbox/CMakeLists.txt]: Have we tested these proficio_sandbox apps?

File: proficio_sandbox/box_builder.cpp [hunk 8]
```diff
@@ -0,0 +1,266 @@
+/**
+ * @file box_builder.cpp
```
Comment (bzenowich) [inline proficio_sandbox/box_builder.cpp]: This app is not included in the Cmakelists.txt, why?

File: proficio_sandbox/configurations/README.txt [hunk 9]
```diff
@@ -0,0 +1,19 @@
+PROFICIO DEVELOPMENT REQUIREMENTS:
```
Comment (bzenowich) [inline proficio_sandbox/configurations/README.txt]: This file is outdated/wrong. It conflicts with the main Readme file.

File: proficio_sandbox/haptic_world.h [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        v_type v;
+        int numMissed;
+        v_type cv;
+
+private:
+        DISALLOW_COPY_AND_ASSIGN(NetworkHaptics);
+};
+
+#endif /* HAPTIC_WORLD_H_ */
+
```
Comment (bzenowich) [inline proficio_sandbox/haptic_world.h]: Where are the proficio_sandbox/images (below) used? Are they necessary?

File: proficio_sandbox/proficio.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/python
```
Comment (bzenowich) [inline proficio_sandbox/proficio.py]: We should consider deleting this, now that we have a more full-featured Proficio launcher.

File: README.txt [hunk 12]
```diff
@@ -4,14 +4,54 @@ Barrett Technology
...
+For Proficio support only:
+
+ - Before running the Proficio in a partifular configuration or immediately after switching the configuration of the proficio type
+
+$ leftConfig
+or
+$ rightConfig
+based on the configuration of the Proficio.
+
+The above bash aliases copies the particular configuration of the proficio from ~/libbarrett/proficio_sandbox/configurations into /etc/barrett directory.
```
Comment (bzenowich) [inline README.txt:64]: We should include instructions on how to create local copies of the configs in ~/.barrett, and info about how libbarrett will always look in ~/.barrett before falling back to /etc/barrett. We should also note that the leftConfig and rightConfig scripts will only update /etc/barrett, not their local configs.

File: CMakeLists.txt [hunk 13]
```diff
@@ -181,7 +180,14 @@ endif()
...
 #     ... others?
-set(config_dir /etc/barrett/)
+set(BARRETT_ETC_PATH /etc/barrett)
+configure_file(
+  ${PROJECT_SOURCE_DIR}/include/barrett/config.h.in 
+  ${PROJECT_BINARY_DIR}/include/barrett/config.h)
+install(FILES ${PROJECT_BINARY_DIR}/include/barrett/config.h 
+  DESTINATION ${CMAKE_INSTALL_PREFIX}/include/barrett)
+
+set(config_dir ${BARRETT_ETC_PATH})
```
Comment (bzenowich) [inline CMakeLists.txt:190]: We should document how to install the configs locally... what is the command required during "make install"?

File: CMakeLists.txt [hunk 14]
```diff
@@ -181,7 +180,14 @@ endif()
 #     products/zerocal.cpp
 #     products/gravitycal.cpp
 #     ... others?
-set(config_dir /etc/barrett/)
+set(BARRETT_ETC_PATH /etc/barrett)
```
Comment (bzenowich) [inline CMakeLists.txt:183]: Have we ever tested the proper operation of local configs stored in ~/.barrett? Have we proven to ourselves that the system will run without the /etc/barrett directory (by deleting /etc/barrett, then testing each example)?

File: proficio_sandbox/box_builder.cpp [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+/** Main Body
+ *
+ *  Main body will record full session data. 
+ */
+template<size_t DOF>
+int wam_main(int argc, char** argv, ProductManager& pm, systems::Wam<DOF>& wam) {
+
+	BARRETT_UNITS_TEMPLATE_TYPEDEFS(DOF);
+
+  //Right ARM
```
Comment (amyblank) [inline proficio_sandbox/box_builder.cpp]: Maybe add a check for right vs left Proficio?

File: proficio_sandbox/configurations/wam3.conf.left [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+
+	joint_velocity_filter:
+	{
+		type = "low_pass";
+		omega_p = (180, 180, 180);
+	};
+
+	joint_position_control:
+	{
+		kp = (1200, 2000, 900);
```
Comment (amyblank) [inline proficio_sandbox/configurations/wam3.conf.left:79]: We should tune these to reduce oscillations. kp=(1200,2000,500) worked better for me.

File: proficio_sandbox/configurations/wam3.conf.right [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+	joint_velocity_filter:
+	{
+		type = "low_pass";
+		omega_p = (180, 180, 180);
+	};
+
+	joint_position_control:
+	{
+		kp = (1200, 2000, 900);
```
Comment (amyblank) [inline proficio_sandbox/configurations/wam3.conf.right:80]: Same as other conf file: We should tune these to reduce oscillations. kp=(1200,2000,500) worked better for me.

File: proficio_sandbox/dual_rehab.cpp [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+		boost::function<void (ProductManager&, systems::Wam<3>&)> wt3);
+template <size_t DOF> void wamThread0(ProductManager& pm0, systems::Wam<DOF>& wam0);
+template <size_t DOF> void wamThread1(ProductManager& pm1, systems::Wam<DOF>& wam1);
+
+
+int main(int argc, char** argv) {
+	// Give us pretty stack-traces when things die
+	installExceptionHandler();
+
+	//ProductManager pm0("/etc/barrett/bus0/bus0.conf");
```
Comment (amyblank) [inline proficio_sandbox/dual_rehab.cpp]: Why is Wam 0 commented out?

File: proficio_sandbox/haptic_world.h [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+        }
+
+protected:
+        virtual void operate() {
+                v = input.getValue();
+
+                {
+                        // send() and recv() cause switches to secondary mode. The socket is
+                        // non-blocking, so this *probably* won't impact the control-loop
+                        // timing that much...
```
Comment (amyblank) [inline proficio_sandbox/haptic_world.h]: Is this something that should be tested? At least the wording of the comment should be changed before the general public can read it.

File: proficio_sandbox/mirror_image.cpp [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+
+		case 't':
+			size_t jointIndex;
+			{
+				size_t jointNumber;
+				std::cout << "\tJoint: ";
+				std::cin >> jointNumber;
+				jointIndex = jointNumber - 1;
+
+				if (jointIndex >= DOF) {
```
Comment (amyblank) [inline proficio_sandbox/mirror_image.cpp]: What if the user enters a negative joint number or a character that is not a number?

File: CMakeLists.txt (hunk 4 above)
Comment (mhariharasudan) [inline CMakeLists.txt:75]: In some linux commands, trailing slash acts on the sub folders while without it acts on the folder itself. I assume that this change includes the include directory itself and all its sub directories while the previous one includes just the sub folders. I am not sure about how important this is since I did not do this. However, I found on the internet that few projects that uses cmake had their include directories listed this way.

File: CMakeLists.txt (hunk 14 above)
Comment (mhariharasudan) [inline CMakeLists.txt:183]: Yes, this has been tested. However, as discussed, we have to manually populate ~/.barrett directory as opposed to copying the files into it using cmake.

File: CMakeLists.txt (hunk 13 above)
Comment (mhariharasudan) [inline CMakeLists.txt:190]: In our previous meeting, we have decided that it is up to the user to populate and maintain the ~/.barrett directory and make install installs the config files to /etc/barrett/

File: README.txt (hunk 6 above)
Comment (mhariharasudan) [inline README.txt:30]: On github, we can change the default branch to any branch. Currently, I have set the default branch to "devel".

File: proficio_sandbox/CMakeLists.txt (hunk 7 above)
Comment (mhariharasudan) [inline proficio_sandbox/CMakeLists.txt]: Since we have not changed any of the sandbox apps, I believe this might come under regression testing. The regression testing will be performed after the code review. I assume lot of examples will fail because of the left/right transformation in the world-to-base transformation. We have to decide whether it is important to modify the sandbox apps to support the Proficio or if it is up to the users, given the limited user base that we have for Proficio.

File: proficio_sandbox/dual_rehab.cpp (hunk 18 above)
Comment (mhariharasudan) [inline proficio_sandbox/dual_rehab.cpp]: This is the same example as dual_rehab_wam.cpp but  believe that this is not yet complete. We can modify this example and get it running. Can you please raise this as an issue, with enhancement tag and assign it to me?

File: proficio_sandbox/haptic_world.h (hunk 19 above)
Comment (mhariharasudan) [inline proficio_sandbox/haptic_world.h]: We may have to discuss about what all files we need to keep. We have a version of haptic world that we are maintaining separately in a different repository.

File: proficio_sandbox/mirror_image.cpp [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+				gainTmp = wam.jpController.getKi();
+				break;
+			case 'd':
+				gainTmp = wam.jpController.getKd();
+				break;
+
+			default:
+				std::cout << "\tBad gain identifier.";
+			}
+			std::cout << gainTmp[jointIndex] << std::endl;
```
Comment (amyblank) [inline proficio_sandbox/mirror_image.cpp]: This is less important, but better to skip this part if the user entered a bad gain identifier.

File: README.txt [hunk 22]
```diff
@@ -4,14 +4,73 @@ Barrett Technology
...
+$ cd libconfig-1.4.5
+$ ./configure && make && sudo make install
+$ cd ../
+$ rm -rf libconfig-1.4.5 libconfig-1.4.5-PATCHED.tar.gz
+
+Download and install libbarrett:
+
+$ cd ~/
+$ git clone https://github.com/BarrettTechnology/libbarrett.git
+$ mv libbarrett/.bash_aliases .
```
Comment (amyblank) [inline README.txt]: Line 31 should be removed because it conflicts with lines 48-54 below.

**Comment by mhariharasudan** [comment]
This went through a code review with @bzenowich @amyblank  and myself. It was tested by GE and was passed on 10/07/2015.
